Performance Insights configuration for aws_db_instance

msvab opened the pull request:
<!--- Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests --->
Fixes #5259

Changes proposed in this pull request:

* Add `performance_insights_enabled` to aws_db_instance
* Add `performance_insights_kms_key_id` to aws_db_instance
* Add `performance_insights_retention_period` to aws_db_instance

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSRDSDBInstance_withPerformanceInsights'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSRDSDBInstance_withPerformanceInsights -timeout 120m
=== RUN   TestAccAWSRDSDBInstance_withPerformanceInsights
=== PAUSE TestAccAWSRDSDBInstance_withPerformanceInsights
=== CONT  TestAccAWSRDSDBInstance_withPerformanceInsights
--- PASS: TestAccAWSRDSDBInstance_withPerformanceInsights (596.38s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	596.419s
```

jalev commented:
👍

msvab commented on aws/resource_aws_db_instance.go:
```diff
@@ -422,6 +422,25 @@ func resourceAwsDbInstance() *schema.Resource {
+			"performance_insights_retention_period": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				Computed: true,
```
This has to be `nil` by default. Because if insights are disabled the SDK doesn't allow any value here.

msvab commented:
Thanks for the detailed feedback, I've implemented all the changes. Let me know if you need any more changes to be done!

msvab commented on aws/resource_aws_db_instance_test.go:
```diff
@@ -4349,3 +4375,44 @@ resource "aws_db_instance" "test" {
 `, rName, rName, rName, rName)
 }
+
+func testAccAWSDBInstancePerformanceInsights(n int) string {
```
This change isn't possible because when insights are disabled, the kms key and retention values have to be empty. So I've used different test configuration to test disabled insights.

jrobison-sb commented:
Any update on if the above changes will be accepted? We'd love to use this once it get released.

bflad commented on aws/resource_aws_db_instance.go:
```diff
@@ -571,6 +588,18 @@ func resourceAwsDbInstanceCreate(d *schema.ResourceData, meta interface{}) error
 		}
 
+		if attr, ok := d.GetOk("performance_insights_enabled"); ok {
+			opts.EnablePerformanceInsights = aws.Bool(attr.(bool))
```
This should have an acceptance test added to ensure its functionality and prevent regressions, e.g. something similar to:

```go
func TestAccAWSDBInstance_ReplicateSourceDb_PerformanceInsightsEnabled(t *testing.T) {
	var dbInstance, sourceDbInstance rds.DBInstance

	rName := acctest.RandomWithPrefix("tf-acc-test")
	kmsKeyResourceName := "aws_kms_key.test"
	sourceResourceName := "aws_db_instance.source"
	resourceName := "aws_db_instance.test"

	resource.ParallelTest(t, resource.TestCase{
		PreCheck:     func() { testAccPreCheck(t) },
		Providers:    testAccProviders,
		CheckDestroy: testAccCheckAWSDBInstanceDestroy,
		Steps: []resource.TestStep{
			{
				Config: testAccAWSDBInstanceConfig_ReplicateSourceDb_PerformanceInsightsEnabled(rName),
				Check: resource.ComposeTestCheckFunc(
					testAccCheckAWSDBInstanceExists(sourceResourceName, &sourceDbInstance),
					testAccCheckAWSDBInstanceExists(resourceName, &dbInstance),
					testAccCheckAWSDBInstanceReplicaAttributes(&sourceDbInstance, &dbInstance),
					resource.TestCheckResourceAttr(resourceName, "performance_insights_enabled", "true"),
					resource.TestCheckResourceAttrPair(resourceName, "performance_insights_kms_key_id", kmsKeyResourceName, "arn"),
					resource.TestCheckResourceAttr(resourceName, "performance_insights_retention_period", "7"),
				),
			},
		},
	})
}

func testAccAWSDBInstanceConfig_ReplicateSourceDb_PerformanceInsightsEnabled(rName string) string {
	return fmt.Sprintf(`
resource "aws_kms_key" "test" {
  description = "Terraform acc test"
  policy      = <<POLICY
{
  "Version": "2012-10-17",
  "Id": "kms-tf-1",
  "Statement": [
    {
      "Sid": "Enable IAM User Permissions",
      "Effect": "Allow",
      "Principal": {
        "AWS": "*"
      },
      "Action": "kms:*",
      "Resource": "*"
    }
  ]
}
POLICY
}

resource "aws_db_instance" "source" {
  allocated_storage       = 5
  backup_retention_period = 1
  engine                  = "mysql"
  engine_version          = "5.6.41"
  identifier              = "%s-source"
  instance_class          = "db.m3.medium"
  password                = "avoid-plaintext-passwords"
  username                = "tfacctest"
  skip_final_snapshot     = true
}

resource "aws_db_instance" "test" {
  identifier                            = %q
  instance_class                        = "${aws_db_instance.source.instance_class}"
  performance_insights_enabled          = true
  performance_insights_kms_key_id       = "${aws_kms_key.test.arn}"
  performance_insights_retention_period = 7
  replicate_source_db                   = "${aws_db_instance.source.id}"
  skip_final_snapshot                   = true
}
`, rName, rName)
}
```

bflad commented on aws/resource_aws_db_instance.go:
```diff
@@ -422,6 +422,23 @@ func resourceAwsDbInstance() *schema.Resource {
 				Optional: true,
 			},
 
+			"performance_insights_enabled": {
```
This attribute (along with the other `performance_insights_*` attributes) are not currently handled during the `RestoreDBInstanceFromDBSnapshot` creation flow so attempting to creating a configuration with PI enabled and `snapshot_identifier` will require two applies of the configuration to converge.

While `RestoreDBInstanceFromDBSnapshot` does not directly support setting `EnablePerformanceInsights`, `PerformanceInsightsKmsKeyId`, and `PerformanceInsightsRetentionPeriod`, we have a framework in place to handle this situation by calling `ModifyDBInstance` immediately after restore, e.g.

```go
// Provided as an example of this logic so similar logic can be added near it
		if attr, ok := d.GetOk("backup_window"); ok {
			modifyDbInstanceInput.PreferredBackupWindow = aws.String(attr.(string))
			requiresModifyDbInstance = true
		}
```

This should have an acceptance test added to ensure its functionality and prevent regressions, e.g. something similar to:

```go
func TestAccAWSDBInstance_SnapshotIdentifier_PerformanceInsightsEnabled(t *testing.T) {
	var dbInstance, sourceDbInstance rds.DBInstance
	var dbSnapshot rds.DBSnapshot

	rName := acctest.RandomWithPrefix("tf-acc-test")
	kmsKeyResourceName := "aws_kms_key.test"
	sourceDbResourceName := "aws_db_instance.source"
	snapshotResourceName := "aws_db_snapshot.test"
	resourceName := "aws_db_instance.test"

	resource.ParallelTest(t, resource.TestCase{
		PreCheck:     func() { testAccPreCheck(t) },
		Providers:    testAccProviders,
		CheckDestroy: testAccCheckAWSDBInstanceDestroy,
		Steps: []resource.TestStep{
			{
				Config: testAccAWSDBInstanceConfig_SnapshotIdentifier_PerformanceInsightsEnabled(rName),
				Check: resource.ComposeTestCheckFunc(
					testAccCheckAWSDBInstanceExists(sourceDbResourceName, &sourceDbInstance),
					testAccCheckDbSnapshotExists(snapshotResourceName, &dbSnapshot),
					testAccCheckAWSDBInstanceExists(resourceName, &dbInstance),
					resource.TestCheckResourceAttr(resourceName, "performance_insights_enabled", "true"),
					resource.TestCheckResourceAttrPair(resourceName, "performance_insights_kms_key_id", kmsKeyResourceName, "arn"),
					resource.TestCheckResourceAttr(resourceName, "performance_insights_retention_period", "7"),
				),
			},
		},
	})
}

func testAccAWSDBInstanceConfig_SnapshotIdentifier_PerformanceInsightsEnabled(rName string) string {
	return fmt.Sprintf(`
resource "aws_kms_key" "test" {
  description = "Terraform acc test"
  policy      = <<POLICY
{
  "Version": "2012-10-17",
  "Id": "kms-tf-1",
  "Statement": [
    {
      "Sid": "Enable IAM User Permissions",
      "Effect": "Allow",
      "Principal": {
        "AWS": "*"
      },
      "Action": "kms:*",
      "Resource": "*"
    }
  ]
}
POLICY
}

resource "aws_db_instance" "source" {
  allocated_storage   = 5
  engine              = "mysql"
  engine_version      = "5.6.41"
  identifier          = "%s-source"
  instance_class      = "db.m3.medium"
  password            = "avoid-plaintext-passwords"
  username            = "tfacctest"
  skip_final_snapshot = true
}

resource "aws_db_snapshot" "test" {
  db_instance_identifier = "${aws_db_instance.source.id}"
  db_snapshot_identifier = %q
}

resource "aws_db_instance" "test" {
  identifier                            = %q
  instance_class                        = "${aws_db_instance.source.instance_class}"
  performance_insights_enabled          = true
  performance_insights_kms_key_id       = "${aws_kms_key.test.arn}"
  performance_insights_retention_period = 7
  snapshot_identifier                   = "${aws_db_snapshot.test.id}"
  skip_final_snapshot                   = true
}
`, rName, rName, rName)
}
```

msvab commented:
Hi @bflad, I've added the extra tests and tried to address the rest of the feedback, but I got a bit stuck and I'm not sure I can finish this on my own since I lack the knowledge of your testing framework.

So once PI is enabled, the KMS Key cannot ever be changed. Even if you disable it and query the API, it will return the key in the response, then if you enable it it will be enabled with the original key. So once KMS Key is set, you can only update `performance_insights_enabled` or `performance_insights_retention_period`. I've updated the tests to describe these scenarios, but failed to make it work. Currently the test fails with ```
=== RUN   TestAccAWSRDSDBInstance_withPerformanceInsights
=== PAUSE TestAccAWSRDSDBInstance_withPerformanceInsights
=== CONT  TestAccAWSRDSDBInstance_withPerformanceInsights
--- FAIL: TestAccAWSRDSDBInstance_withPerformanceInsights (665.50s)
    testing.go:538: Step 1 error: After applying this step, the plan was not empty:
        
        DIFF:
        
        UPDATE: aws_db_instance.bar
          performance_insights_kms_key_id: "arn:aws:kms:eu-central-1:624670762265:key/bf1ee5ca-2491-47e8-8e88-db54a051d5d5" => ""
        
        STATE:
        
        aws_db_instance.bar:
          ID = tf-db-test-7067100361891765948
          provider = provider.aws
          address = tf-db-test-7067100361891765948.cekwdbllig4s.eu-central-1.rds.amazonaws.com
          allocated_storage = 5
          arn = arn:aws:rds:eu-central-1:624670762265:db:tf-db-test-7067100361891765948
          auto_minor_version_upgrade = true
          availability_zone = eu-central-1a
          backup_retention_period = 0
          backup_window = 20:24-20:54
          ca_cert_identifier = rds-ca-2015
          copy_tags_to_snapshot = false
          db_subnet_group_name = default
          deletion_protection = false
          domain = 
          domain_iam_role_name = 
          enabled_cloudwatch_logs_exports.# = 0
          endpoint = tf-db-test-7067100361891765948.cekwdbllig4s.eu-central-1.rds.amazonaws.com:3306
          engine = mysql
          engine_version = 5.6.41
          hosted_zone_id = Z1RLNUO7B9Q6NB
          iam_database_authentication_enabled = false
          identifier = tf-db-test-7067100361891765948
          instance_class = db.m3.medium
          iops = 0
          kms_key_id = 
          license_model = general-public-license
          maintenance_window = wed:01:33-wed:02:03
          monitoring_interval = 0
          multi_az = false
          name = mydb
          option_group_name = default:mysql-5-6
          parameter_group_name = default.mysql5.6
          password = mustbeeightcharaters
          performance_insights_enabled = false
          performance_insights_kms_key_id = arn:aws:kms:eu-central-1:624670762265:key/bf1ee5ca-2491-47e8-8e88-db54a051d5d5
          performance_insights_retention_period = 0
          port = 3306
          publicly_accessible = false
          replicas.# = 0
          replicate_source_db = 
          resource_id = db-IXOXIOGYPQOQ2APEYIY5HDYLUM
          security_group_names.# = 0
          skip_final_snapshot = true
          status = available
          storage_encrypted = false
          storage_type = gp2
          tags.% = 0
          timezone = 
          username = foo
          vpc_security_group_ids.# = 1
          vpc_security_group_ids.1630437501 = sg-40f0442d
        aws_kms_key.foo:
          ID = bf1ee5ca-2491-47e8-8e88-db54a051d5d5
          provider = provider.aws
          arn = arn:aws:kms:eu-central-1:624670762265:key/bf1ee5ca-2491-47e8-8e88-db54a051d5d5
          description = Terraform acc test 7067100361891765948
          enable_key_rotation = false
          is_enabled = true
          key_id = bf1ee5ca-2491-47e8-8e88-db54a051d5d5
          key_usage = ENCRYPT_DECRYPT
          policy = {"Id":"kms-tf-1","Statement":[{"Action":"kms:*","Effect":"Allow","Principal":{"AWS":"*"},"Resource":"*","Sid":"Enable IAM User Permissions"}],"Version":"2012-10-17"}
          tags.% = 0
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	665.548s
make: *** [testacc] Error 1```

analytically commented:
FYI https://aws.amazon.com/about-aws/whats-new/2019/02/performance-insights-supports-T2-and-T3/

Sytten commented:
Any ETA for a merge?

RossHammer commented:
@msvab I looked at your PR and found there were different things that needed changing to get the tests passing. It seems there are a couple things that are misconfigured to get it working, here is the diff I found that gets everything operating nicely.

```
--- a/aws/resource_aws_db_instance.go
+++ b/aws/resource_aws_db_instance.go
@@ -433,11 +433,13 @@ func resourceAwsDbInstance() *schema.Resource {
 			"performance_insights_kms_key_id": {
 				Type:         schema.TypeString,
 				Optional:     true,
+				Computed:     true,
 				ValidateFunc: validateArn,
 			},
 
 			"performance_insights_retention_period": {
 				Type:     schema.TypeInt,
+				Computed: true,
 				Optional: true,
 			},
 
@@ -1550,7 +1552,7 @@ func resourceAwsDbInstanceUpdate(d *schema.ResourceData, meta interface{}) error
 		requestUpdate = true
 	}
 
-	if d.Get("performance_insights_enabled").(bool) && d.HasChange("performance_insights_kms_key_id") {
+	if d.Get("performance_insights_enabled").(bool) && (d.HasChange("performance_insights_kms_key_id") || d.HasChange("performance_insights_enabled")) {
 		d.SetPartial("performance_insights_kms_key_id")
 		req.PerformanceInsightsKMSKeyId = aws.String(d.Get("performance_insights_kms_key_id").(string))
 		requestUpdate = true
```

msvab commented:
Thanks @RossHammer! I've updated the code and the tests are passing now.

bflad commented:
This has been released in [version 2.12.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#2120-may-24-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
